climada/hazard/base.py

```diff
@@ -1101,41 +1112,6 @@ def read_hdf5(self, file_name):
 
         hf_data.close()
 
-    def concatenate(self, haz_src, append=False):
```

Comment:
I checked through the repo (and the papers repo) for anywhere that still uses this. I think the emulator tutorial still does and that's it

Reply:
If you haven't already, we should mention that this is being removed in the release notes, because it could break things on branches and forks we haven't checked/can't see

Reply:
That is a good point. It will be mentioned in the release notes. Should I make a deprecation notice for `self.concatenate` maybe?

Reply:
Also, what do you think about the naming:

There is now `hazard.concatenate_hazard`, `hazard.extend` and `hazard.append`. I think this is too much.

Reply:
@emanuel-schmid Any thoughts on the naming and the note about removing the method `self.concatenate` ?

Reply:
about the deprecation notice - what do you mean? keep the method and log a warning like with the all but removed `IF_xy methods`? Would that even be an option in this case?

Reply:
@emanuel-schmid isn't the analogy with `list` fairly close? I see `append` and `extend` as operations that take either a single element or an iterable of elements and combine them into `self`.

Reply:
@chahank I think this is probably too many methods, yes. With `hazard.concatenate` we've a few options:
- Remove it, as we're doing now, and hope nothing breaks in places we can't see
- Add a deprecation notice and remove it in a future release
- Just add a deprecation notice
The second of these is probably 'best practice' but our user base is not so big we couldn't do the first.

With `extend`, `append`, and `concatenate_hazard`, it is a lot. I think it's worth pointing out similar methods in the docstrings with a 'See also...', so users are aware of them. I'll add these as suggestions now.

We could also think about deprecating `hazard.append` because `hazard.extend` is almost always better. A deprecation notice would be annoying to anyone using the method, but in the long run it would encourage (slightly) better code ... I'm not sure it's worth our (and our users') time, though.

Finally, should we add the parameters `centroids` and `threshold` to `hazard.append` for consistency? It won't break backward compatibility.

Reply:
> @emanuel-schmid isn't the analogy with `list` fairly close? I see `append` and `extend` as operations that take either a single element or an iterable of elements and combine them into `self`.

Agreed, that's a valid perspective if we accept that Hazard has a dual nature, being a list and an object at the same time.

Reply:
> With `extend`, `append`, and `concatenate_hazard`, it is a lot. I think it's worth pointing out similar methods in the docstrings with a 'See also...', so users are aware of them. I'll add these as suggestions now.
> 
> We could also think about deprecating `hazard.append` because `hazard.extend` is almost always better. A deprecation notice would be annoying to anyone using the method, but in the long run it would encourage (slightly) better code ... I'm not sure it's worth our (and our users') time, though.
> 

Both methods are very stateful and thus - to some extent - calling for trouble. Me, personally, I wouldn't even be sad if we discarded both of them. In any case we could think about encouraging folks to use `concatenate_hazard`.

Reply:
> Finally, should we add the parameters centroids and threshold to hazard.append for consistency? It won't break backward compatibility.
> 
👍

Reply:
>Agreed, that's a valid perspective if we accept that Hazard has a dual nature, being a list and an object at the same time.
It kind of is. It is an object Hazard and a list of events.

Reply:
>Me, personally, I wouldn't even be sad if we discarded both of them. In any case we could think about encouraging folks to use concatenate_hazard.

The reason we have an extend method is that Hazard objects can get large in memory, and thus it would be nice to have a method to append without having afterwards twice the object.

Reply:
I see your point, but the argument is somewhat mitigated by the fact that the critical maximum memory consumption is not a bit smaller, for `conatenate_hazard` is called in any case. The difference is that with `extend` the garbage collector is automatically called for action and with `concatenate_hazard` one would have to free memory explicitly...

Reply:
Yes, I see. I mean, having the garbage collector to free memory is also a plus, but not as important as the critical maximum memory.